airflow/providers/google/cloud/hooks/compute_ssh.py

```diff
@@ -0,0 +1,250 @@
+
+        client.connect(hostname=hostname,
+                       username=username,
+                       key_filename=private_key_file)
+        return client
+
+    def get_service_account(self) -> str:
+        """
+        Get the service account id
+        """
+        account: str = self._get_credentials().service_account_email
+        if not account.startswith('users/'):
+            account = 'users/' + account
+        return account
+
+    def get_username(self) -> str:
+        """
+        Using the OS Login API, get the POSIX username from the login profile
+        for the service account.
+
+        """
+        conn = self.get_conn()
+        account = self.get_service_account()
+        profile = conn.users().getLoginProfile(name=account).execute_in_compute()  # pylint: disable=no-member
+        username: str = profile.get('posixAccounts')[0].get('username')
+        return username
+
+    @GoogleBaseHook.fallback_to_default_project_id
+    def get_hostname(self, project_id: Optional[str], instance: str, zone: str,
+                     hostname: Optional[str] = None) -> str:
+        """
+        Create the target instance hostname
+
+        :param project_id: The project ID of the instance
+        :type project_id: Optional[str]
+        :param instance: The name of the target instance to connect to
+        :type instance: str
+        :param zone: The zone of the target instance
+        :param zone: str
+        :param hostname: The hostname of the target instance. If not provided,
+            it is generated from zone, instance name and project id
+        :type hostname: str
+        """
+        if not hostname:
+            hostname = f'{instance}.{zone}.c.{project_id}.internal'
+        return hostname
+
+    def execute_local(self, cmd: Union[str, List[str]]) -> None:
+        """
+        Executes command in local machine
+
+        :param cmd: The command to execute
+        :type cmd: Union[str, List[str]]
+        """
+        self.log.info('Executing command: %s', str(cmd))
+        process = subprocess.Popen(cmd)
+        output = process.communicate()[0]
+        return_code = process.returncode
+        if return_code:
+            # Error
+            raise AirflowException(f'The Command {cmd} returned error status {return_code}')
+        if output:
+            self.log.info(output)
+
+    def create_ssh_key(self,
+                       account: str,
+                       expire_time: int = 300) -> str:
+        """
+        Generate a temporary SSH key and apply it to the specified account
+
+        :param account: The service account id
+        :type account: str
+        :param expire_time: The maximum amount of time before the private key is expired
+        :type expire_time: int
+        """
+        try:
+            self.log.info("Generating ssh keys...")
+            private_key_file = '/tmp/key-' + str(uuid.uuid4())
```

Comment:
```python
def generate_key_string(pkey: paramiko.PKey):
    key_fh = StringIO()
    pkey.write_private_key(key_fh)
    key_fh.seek(0)
    key_str = key_fh.read()
    return key_str
```
You can use this code to prepare SSH key without writing to disk.

Reply:
You can also read key from .key attribute and serialize it to string.
https://cryptography.io/en/latest/hazmat/primitives/asymmetric/rsa/#key-serialization

Reply:
I have tried several ways to use paramiko with StringIO but I always get authentication failure. Using the oslogin and shell command, the system test succeeds but if oslogin is switched off, It doesn't work as expected. 

Another issue is the account username, when I check the server, oslogin is using service account unique ID which is a bunch of numbers. Was expecting it to use the part of email before '@' but it uses the unique ID.

Reply:
Paramiko is now working. Oslogin works but managing key through instant metadata still doesn't work. The key is set correctly as advised [here](https://cloud.google.com/compute/docs/instances/adding-removing-ssh-keys#instance-only) but am still yet to figure out why it's not working